frame/staking/src/offchain_election.rs

```diff
@@ -155,19 +163,180 @@ pub fn get_balancing_iters<T: Trait>() -> usize {
 	}
 }
 
+/// Find the maximum `len` that a compact can have in order to fit into the block weight.
+///
+/// This only returns a value between zero and `size.nominators`.
+pub fn maximum_compact_len<W: crate::WeightInfo>(
+	winners_len: u32,
+	size: ElectionSize,
+	max_weight: Weight,
+) -> u32 {
+	use sp_std::cmp::Ordering;
+
+	if size.nominators < 1 {
+		return size.nominators;
+	}
+
+	let max_voters = size.nominators.max(1);
+	let mut voters = max_voters;
+
+	// helper closures.
+	let weight_with = |voters: u32| -> Weight {
+		W::submit_solution_better(
+			size.validators.into(),
+			size.nominators.into(),
+			voters,
+			winners_len,
+		)
+	};
+
+	let next_voters = |current_weight: Weight, voters: u32, step: u32| -> Result<u32, ()> {
+		match current_weight.cmp(&max_weight) {
+			Ordering::Less => {
+				let next_voters = voters.checked_add(step);
+				match next_voters {
+					Some(voters) if voters < max_voters => Ok(voters),
+					_ => Err(()),
+				}
+			},
+			Ordering::Greater => voters.checked_sub(step).ok_or(()),
+			Ordering::Equal => Ok(voters),
+		}
+	};
+
+	// First binary-search the right amount of voters
+	let mut step = voters / 2;
+	let mut current_weight = weight_with(voters);
+	while step > 0 {
+		match next_voters(current_weight, voters, step) {
+			// proceed with the binary search
+			Ok(next) if next != voters => {
+				voters = next;
+			},
+			// we are out of bounds, break out of the loop.
+			Err(()) => {
+				break;
+			},
+			// we found the right value - early exit the function.
+			Ok(next) => return next
+		}
+		step = step / 2;
+		current_weight = weight_with(voters);
+	}
+
+
+	// Time to finish.
+	// We might have reduced less than expected due to rounding error. Increase one last time if we
+	// have any room left, the reduce until we are sure we are below limit.
+	while weight_with(voters + 1) < max_weight {
```

Comment:
:rofl: that's broken again, we should either be using the `next_voters` closure, to make sure we don't overflow or add `voters + 1< max_voters` to the condition.
If you set `max_weight` to ♾, this loop will probably overflow. If `weight_with(max_voters) << max_weight` this loop will unnecessarily turn multiple times just to have the result be bound again right before returning.

```rust
let prev_voters = 0;
loop {
  let current_weight = weight_with(voters);
  let next_voters = next_voters(current_weight, voters, 1);
  if next_voters == prev_voters || next_voters = voters { break; } // we started spinning
  prev_voters = voters;
  voters = next_voters;
}
```
The loop above should do roughly the same, but I agree it's not really prettier. I didn't cater for the case, where we start alternating between two values around `max_weight` when proposing the initial version. Including this check makes the loop uglier :)

Reply:
>  If weight_with(max_voters) << max_weight this loop will unnecessarily turn multiple times just to have the result be bound again right before returning.

This cannot happen because we first execute the binary part. So we are actually at most a few steps away from the final value. 

Yes so I will add the condition to prevent overflow and let's forget this 😁